src/rules/azureResourceGraph.ts

```diff
@@ -77,35 +101,155 @@ export class ResourceGraphRule implements BaseRule<ResourceGraphTarget> {
       throw Error("Invalid Query. All queries must start with '<tableName> |'");
     } else {
-      const initalTable = this.query.slice(0, firstPipeIndex - 1);
-      const queryEnding = this.query.slice(firstPipeIndex);
+      const initalTable = this.evaluation.query.slice(0, firstPipeIndex - 1);
+      const queryEnding = this.evaluation.query.slice(firstPipeIndex);
       return `${initalTable} ${groupQuery} ${queryEnding}`;
     }
   }
 
-  toScanResult(response: ResourceGraphModels.ResourcesResponse): ScanResult {
+  async sendRequest(target: ResourceGraphTarget, resourceId: string) {
+    if (!isRequestEvaluation(this.evaluation)) {
+      throw Error('A valid request evalutation was not found');
+    }
+    const token = await target.credential.getToken(
+      'https://graph.microsoft.com/.default'
+    );
+    const resourceManagementClient = await this.getResourceManagmentClient(
+      resourceId
+    );
+    const options = {
+      url: await this.getRequestUrl(resourceId, resourceManagementClient),
+      method: this.evaluation.request.httpMethod as HttpMethods,
+      headers: {
+        Authorization: `Bearer ${token?.token}`,
+        'Content-Type': 'application/json',
+      },
+    };
+    let response = await resourceManagementClient.sendRequest(options);
+    // if the response returns an error because the default api verison is invalid, then parse the error message to retrieve a valid one and try again
+    if (
+      response.status === 400 &&
+      response.parsedBody.error.code === 'NoRegisteredProviderFound'
+    ) {
+      const apiVersion = this.getApiVersionFromError(
+        response.parsedBody.error.message
+      );
+      options.url = await this.getRequestUrl(
+        resourceId,
+        resourceManagementClient,
+        apiVersion
+      );
+      response = await resourceManagementClient.sendRequest(options);
+    }
+    return response;
+  }
+
+  async getResourceManagmentClient(resourceId: string) {
+    const subscriptionId = this.getElementFromId('subscription', resourceId);
+    return await new ResourceManagementClient(
+      new AzureIdentityCredentialAdapter(new DefaultAzureCredential()),
+      subscriptionId
+    );
+  }
+
+  getApiVersionFromError(errorMessage: string) {
+    // splits message so the initial api version that failed is not included in the regex match.
+    const splitMessage = errorMessage.split('The supported api-versions are ');
+    const versions = splitMessage[1];
+    const regExps = [
+      new RegExp(/\d\d\d\d-\d\d-\d\d-preview/g),
+      new RegExp(/\d\d\d\d-\d\d-\d\d/g),
+    ];
+    for (const r of regExps) {
+      // first looks for any preview api versions, then looks for latest if preview is not found.
+      const apiVersions = versions.match(r);
+      if (apiVersions) {
+        return apiVersions[apiVersions.length - 1];
+      }
+    }
+    throw Error('Unable to find a valid api version');
+  }
```

Comment:
I added this method because I was previously trying to get the latest api version in the getDefaultApiVersion method. It was coming back with api versions but when I tried the latest, I got an error saying it was not valid for that region (not sure as to why, I need to look into this further).  I changed the getDefaultApiVersion method to retrieve the default api version instead and haven't had an issue yet. I left this method in the code in case there is a problem with the default version, although I'm not sure this is entirely necessary.

Reply:
I figured out what the problem was. I was trying to get the latest version for Microsoft.Eventhub/namespaces when I should have been trying to get Microsoft.Eventhub/namespaces/networkRuleSets. This is an issue because it will be difficult to figure out which resourceType to use. 

I thought using the request.operation value would be the way we could accomplish this. However, in the Event Hub example, the request.operation  is "networkRuleSets/default". Adding the "default" causes it to fail and therefore relys on the error message to save it. 

We could rely on the error message to save us for now but we should figure out a way  to accomplish this in a more stable way.

I need to look further into what the default api version is and if that can be used across a providers resource types

Reply:
I've refactored this method since commenting but the issue still remains